pandas/core/arrays/period.py

```diff
@@ -539,14 +769,154 @@ def _check_timedeltalike_freq_compat(self, other):
+            return self.asfreq(dtype.freq)
+        else:
+            return np.asarray(self, dtype=dtype)
+
+    def _item(self):
+        if len(self) == 1:
+            # IndexOpsMixin will catch and re-raise IndexErrors
+            return Period._from_ordinal(self.values[0], self.freq)
+        else:
+            raise ValueError('can only convert an array of size 1 to a '
+                             'Python scalar')
+
+    @property
+    def flags(self):
+        # TODO: remove
+        # We need this since reduction.SeriesBinGrouper uses values.flags
+        # Ideally, we wouldn't be passing objects down there in the first
+        # place.
+        return self._ndarray_values.flags
+
 
-PeriodArrayMixin._add_comparison_ops()
-PeriodArrayMixin._add_datetimelike_methods()
+
+PeriodArray._add_comparison_ops()
+PeriodArray._add_datetimelike_methods()
 
 
 # -------------------------------------------------------------------
 # Constructor Helpers
 
+def period_array(data, freq=None, ordinal=None, copy=False):
+    # type: (Sequence[Optional[Period]], Optional[Tick]) -> PeriodArray
+    """
+    Construct a new PeriodArray from a sequence of Period scalars.
+
+    Parameters
+    ----------
+    data : Sequence of Period objects
+        A sequence of Period objects. These are required to all have
+        the same ``freq.`` Missing values can be indicated by ``None``
+        or ``pandas.NaT``.
+    freq : str, Tick, or Offset
+        The frequency of every element of the array. This can be specified
+        to avoid inferring the `freq` from `data`.
+    copy : bool, default False
+        Whether to ensure a copy of the data is made.
+
+    Returns
+    -------
+    PeriodArray
+
+    See Also
+    --------
+    PeriodArray
+    pandas.PeriodIndex
+
+    Examples
+    --------
+    >>> period_array([pd.Period('2017', freq='A'),
+    ...               pd.Period('2018', freq='A')])
+    <PeriodArray>
+    ['2017', '2018']
+    Length: 2, dtype: period[A-DEC]
+
+    >>> period_array([pd.Period('2017', freq='A'),
+    ...               pd.Period('2018', freq='A'),
+    ...               pd.NaT])
+    <PeriodArray>
+    ['2017', '2018', 'NaT']
+    Length: 3, dtype: period[A-DEC]
+    """
+
+    if data is None and ordinal is None:
+        raise ValueError("range!")
+    elif data is None:
+        data = np.asarray(ordinal, dtype=np.int64)
+        if copy:
+            data = data.copy()
+        return PeriodArray(data, freq=freq)
+    else:
+        if isinstance(data, (ABCPeriodIndex, ABCSeries, PeriodArray)):
+            return PeriodArray(data, freq)
+        elif is_datetime64_dtype(data):
+            return PeriodArray._from_datetime64(data, freq)
+
+        # other iterable of some kind
+        if not isinstance(data, (np.ndarray, list, tuple)):
+            data = list(data)
+
+    if freq:
+        dtype = PeriodDtype(freq)
+    else:
+        dtype = None
+
+    if lib.infer_dtype(data) == 'floating' and len(data) > 0:
```

Comment:
this is quite fast for the vast majority of things, only object dtype is fully inferred

Reply:
Yeah, we're going to have objects here typically though.

For a length 1,000 array of period objects we spend 70us in `lib.infer_dtype` and 330us in the actual constructor, so about 20% of the time is just to raise this error message. I refactored it a bit to just check for `is_float_dtype`.